File: src/ptr/tagged_pointer.zig
const AddressableSize = u49;

pub const TaggedPointer = packed struct(u64) {
    _ptr: AddressableSize,
    data: Tag,

    pub const Tag = u15;

    pub inline fn init(ptr: anytype, data: Tag) TaggedPointer {
        const Ptr = @TypeOf(ptr);
        if (comptime Ptr == @TypeOf(null)) {
            return .{ ._ptr = 0, .data = data };
        }

        if (comptime @typeInfo(Ptr) != .pointer and Ptr != ?*anyopaque) {
            @compileError(@typeName(Ptr) ++ " must be a ptr, received: " ++ @tagName(@typeInfo(Ptr)));
        }

        const address = @intFromPtr(ptr);

        return TaggedPointer{
            ._ptr = @as(AddressableSize, @truncate(address)),
            .data = data,
        };
    }

    pub inline fn get(this: TaggedPointer, comptime Type: type) *Type {
        return @as(*Type, @ptrFromInt(@as(usize, @intCast(this._ptr))));
    }

    pub inline fn from(val: anytype) TaggedPointer {
        const ValueType = @TypeOf(val);
        return switch (ValueType) {
            f64, i64, u64 => @bitCast(val),
            ?*anyopaque, *anyopaque => @bitCast(@intFromPtr(val)),
            else => @compileError("Unsupported type: " ++ @typeName(ValueType)),
        };
    }

    pub inline fn to(this: TaggedPointer) *anyopaque {
        return @ptrFromInt(@as(u64, @bitCast(this)));
    }
};

const TypeMapT = struct {
    value: TaggedPointer.Tag,
    ty: type,
    name: []const u8,
};
pub fn TypeMap(comptime Types: anytype) type {
    return [Types.len]TypeMapT;
}

pub fn TagTypeEnumWithTypeMap(comptime Types: anytype) struct {
    tag_type: type,
    ty_map: TypeMap(Types),
} {
    var typeMap: TypeMap(Types) = undefined;
    var enumFields: [Types.len]std.builtin.Type.EnumField = undefined;

    @memset(&enumFields, std.mem.zeroes(std.builtin.Type.EnumField));
    @memset(&typeMap, TypeMapT{ .value = 0, .ty = void, .name = "" });

    inline for (Types, 0..) |field, i| {
        const name = comptime @typeName(field);
        enumFields[i] = .{
            .name = name,
            .value = 1024 - i,
        };
        typeMap[i] = .{ .value = 1024 - i, .ty = field, .name = name };
    }

    return .{
        .tag_type = @Type(.{
            .@"enum" = .{
                .tag_type = TaggedPointer.Tag,
                .fields = &enumFields,
                .decls = &.{},
                .is_exhaustive = false,
            },
        }),
        .ty_map = typeMap,
    };
}

pub fn TaggedPointerUnion(comptime Types: anytype) type {
    const result = TagTypeEnumWithTypeMap(Types);

    const TagType: type = result.tag_type;

    return struct {
        pub const Tag = TagType;
        pub const TagInt = TaggedPointer.Tag;
        pub const type_map: TypeMap(Types) = result.ty_map;
        repr: TaggedPointer,

        pub const Null: @This() = .{ .repr = .{ ._ptr = 0, .data = 0 } };

        pub fn clear(this: *@This()) void {
            this.* = Null;
        }

        pub fn typeFromTag(comptime the_tag: comptime_int) type {
            for (type_map) |entry| {
                if (entry.value == the_tag) return entry.ty;
            }
            @compileError("Unknown tag: " ++ the_tag);
        }

        pub fn typeNameFromTag(the_tag: TagInt) ?[]const u8 {
            inline for (type_map) |entry| {
                if (entry.value == the_tag) return entry.name;
            }
            return null;
        }

        pub fn typeName(this: This) ?[]const u8 {
            return @tagName(this.tag());
        }

        const This = @This();
        pub fn assert_type(comptime Type: type) void {
            const name = comptime @typeName(Type);
            if (!comptime @hasField(Tag, name)) {
                @compileError("TaggedPointerUnion does not have " ++ name ++ ".");
            }
        }
        pub inline fn get(this: This, comptime Type: type) ?*Type {
            comptime assert_type(Type);

            return if (this.is(Type)) this.as(Type) else null;
        }

        pub inline fn tag(this: This) TagType {
            return @as(TagType, @enumFromInt(this.repr.data));
        }

        pub fn case(comptime Type: type) Tag {
            return @field(Tag, bun.meta.typeBaseName(@typeName(Type)));
        }

        /// unsafely cast a tagged pointer to a specific type, without checking that it's really that type
        pub inline fn as(this: This, comptime Type: type) *Type {
            comptime assert_type(Type);
            return this.repr.get(Type);
        }

        pub inline fn setUintptr(this: *This, value: AddressableSize) void {
            this.repr._ptr = value;
        }

        pub inline fn asUintptr(this: This) AddressableSize {
            return this.repr._ptr;
        }

        pub inline fn is(this: This, comptime Type: type) bool {
            comptime assert_type(Type);
            return this.repr.data == comptime @intFromEnum(@field(Tag, @typeName(Type)));
        }

        pub fn set(this: *@This(), _ptr: anytype) void {
            this.* = @This().init(_ptr);
        }

        pub inline fn isValidPtr(_ptr: ?*anyopaque) bool {
            return This.isValid(This.from(_ptr));
        }

        pub inline fn isValid(this: This) bool {
            return switch (this.repr.data) {
                @intFromEnum(
                    @field(Tag, @typeName(Types[Types.len - 1])),
                )...@intFromEnum(
                    @field(Tag, @typeName(Types[0])),
                ) => true,
                else => false,
            };
        }

        pub inline fn from(_ptr: ?*anyopaque) This {
            return This{ .repr = TaggedPointer.from(_ptr) };
        }

        pub inline fn ptr(this: This) *anyopaque {
            return this.repr.to();
        }

        pub inline fn ptrUnsafe(this: This) *anyopaque {
            @setRuntimeSafety(false);
            return this.repr.to();
        }

        pub inline fn init(_ptr: anytype) @This() {
            const tyinfo = @typeInfo(@TypeOf(_ptr));
            if (tyinfo != .pointer) @compileError("Only pass pointers to TaggedPointerUnion.init(), you gave us a: " ++ @typeName(@TypeOf(_ptr)));

            const Type = std.meta.Child(@TypeOf(_ptr));
            return initWithType(Type, _ptr);
        }

        pub inline fn initWithType(comptime Type: type, _ptr: anytype) @This() {
            const tyinfo = @typeInfo(@TypeOf(_ptr));
            if (tyinfo != .pointer) @compileError("Only pass pointers to TaggedPointerUnion.init(), you gave us a: " ++ @typeName(@TypeOf(_ptr)));
            const name = comptime @typeName(Type);

            // there will be a compiler error if the passed in type doesn't exist in the enum
            return This{ .repr = TaggedPointer.init(_ptr, @intFromEnum(@field(Tag, name))) };
        }

        pub inline fn isNull(this: This) bool {
            return this.repr._ptr == 0;
        }

        pub inline fn call(this: This, comptime fn_name: []const u8, args_without_this: anytype, comptime Ret: type) Ret {
            inline for (type_map) |entry| {
                if (this.repr.data == entry.value) {
                    const pointer = this.as(entry.ty);
                    const func = &@field(entry.ty, fn_name);
                    const args = brk: {
                        var args: std.meta.ArgsTuple(@TypeOf(@field(entry.ty, fn_name))) = undefined;
                        args[0] = pointer;

                        inline for (args_without_this, 1..) |a, i| {
                            args[i] = a;
                        }

                        break :brk args;
                    };
                    return @call(.auto, func, args);
                }
            }
            @panic("Invalid tag");
        }
    };
}

const bun = @import("bun");
const std = @import("std");

File: src/ptr/tagged_pointer.rs
use core::ffi::c_void;
use core::marker::PhantomData;

/// Logically a u49. Rust has no native u49, so we carry it in a u64 and mask.
pub(crate) type AddressableSize = u64;

/// `TaggedPtr::Tag` — logically u15, carried in u16. (Inherent assoc types are nightly; hoisted here.)
pub type TagType = u16;
/// Bit layout:
///   bits  0..49 → `_ptr`
///   bits 49..64 → `data`
#[repr(transparent)]
#[derive(Copy, Clone, Eq, PartialEq, Hash)]
pub struct TaggedPtr(u64);

impl TaggedPtr {
    /// Logically a u15.
    // pub type Tag = u16; → hoisted to module-level TagType (inherent assoc types are nightly-only)

    const ADDR_BITS: u32 = 49;
    const ADDR_MASK: u64 = (1u64 << Self::ADDR_BITS) - 1;
    const TAG_MASK: u16 = (1u16 << 15) - 1;

    #[inline]
    pub fn init<T>(ptr: *const T, data: TagType) -> TaggedPtr {
        // A null `*const T` yields address 0 below.
        let address = ptr as usize;
        TaggedPtr(
            (address as u64 & Self::ADDR_MASK) // @truncate to u49
                | ((data as u64 & Self::TAG_MASK as u64) << Self::ADDR_BITS),
        )
    }

    #[inline]
    fn ptr_bits(self) -> AddressableSize {
        self.0 & Self::ADDR_MASK
    }

    #[inline]
    pub fn data(self) -> TagType {
        (self.0 >> Self::ADDR_BITS) as TagType
    }

    #[inline]
    pub fn get<Type>(self) -> *mut Type {
        // SAFETY: caller asserts the stored address points to a live `Type`.
        self.ptr_bits() as usize as *mut Type
    }

    #[inline]
    pub fn to(self) -> *mut c_void {
        // Note: includes the tag bits in the high word. This is intentional:
        // round-tripping through `*anyopaque` preserves the tag.
        self.0 as usize as *mut c_void
    }
}

impl From<u64> for TaggedPtr {
    #[inline]
    fn from(val: u64) -> Self {
        TaggedPtr(val)
    }
}
impl From<*mut c_void> for TaggedPtr {
    #[inline]
    fn from(val: *mut c_void) -> Self {
        TaggedPtr(val as usize as u64)
    }
}
impl From<Option<*mut c_void>> for TaggedPtr {
    #[inline]
    fn from(val: Option<*mut c_void>) -> Self {
        TaggedPtr(val.map_or(0, |p| p as usize as u64))
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// TaggedPointerUnion
// ─────────────────────────────────────────────────────────────────────────────
//
// Modeled with two traits:
//
//   - `TypeList`      : implemented for the tuple `(T1, T2, ...)`; carries the
//                       tag range.
//   - `UnionMember<L>`: implemented for each `Ti` against its list `L`; carries
//                       that type's tag value (1024 - index).
//
// Membership checks become a `T: UnionMember<Ts>` bound — the compile error is
// the trait-bound failure.
//
// Tag values are assigned as `1024 - i` for index `i`. We keep the raw u15
// and expose `tag()` as the integer (callers that need an enum can define one
// per instantiation).

/// Implemented for the tuple of types passed to `TaggedPtrUnion<(...)>`.
pub trait TypeList {
    /// `@intFromEnum(@field(Tag, @typeName(Types[Types.len - 1])))` = 1024 - (LEN-1)
    const MIN_TAG: TagType;
    /// `@intFromEnum(@field(Tag, @typeName(Types[0])))` = 1024
    const MAX_TAG: TagType = 1024;
}

/// `T: UnionMember<Ts>` ⇔ `T` is one of the types in the list `Ts`.
pub trait UnionMember<Ts: TypeList> {
    const TAG: TagType;
}

#[repr(transparent)]
pub struct TaggedPtrUnion<Ts: TypeList> {
    repr: TaggedPtr,
    _types: PhantomData<Ts>,
}

impl<Ts: TypeList> Clone for TaggedPtrUnion<Ts> {
    fn clone(&self) -> Self {
        *self
    }
}
impl<Ts: TypeList> Copy for TaggedPtrUnion<Ts> {}

impl<Ts: TypeList> TaggedPtrUnion<Ts> {
    // pub type TagInt → use module-level TagType

    #[inline]
    pub fn get<Type: UnionMember<Ts>>(self) -> Option<*mut Type> {
        if self.is::<Type>() {
            Some(self.as_unchecked::<Type>())
        } else {
            None
        }
    }

    #[inline]
    pub fn tag(self) -> TagType {
        self.repr.data()
    }

    /// The tag constant for `Type`.
    #[inline]
    pub const fn case<Type: UnionMember<Ts>>() -> TagType {
        Type::TAG
    }

    /// unsafely cast a tagged pointer to a specific type, without checking that it's really that type
    // `as` is a Rust keyword, hence `as_unchecked`.
    #[inline]
    pub fn as_unchecked<Type: UnionMember<Ts>>(self) -> *mut Type {
        self.repr.get::<Type>()
    }

    #[inline]
    pub fn as_uintptr(self) -> AddressableSize {
        self.repr.ptr_bits()
    }

    #[inline]
    pub fn is<Type: UnionMember<Ts>>(self) -> bool {
        self.repr.data() == Type::TAG
    }

    #[inline]
    pub fn is_valid(self) -> bool {
        // switch (data) { MIN_TAG...MAX_TAG => true, else => false }
        let d = self.repr.data();
        d >= Ts::MIN_TAG && d <= Ts::MAX_TAG
    }

    #[inline]
    pub fn from(ptr: Option<*mut c_void>) -> Self {
        Self {
            repr: TaggedPtr::from(ptr),
            _types: PhantomData,
        }
    }

    #[inline]
    pub fn ptr(self) -> *mut c_void {
        self.repr.to()
    }

    #[inline]
    pub fn init<Type: UnionMember<Ts>>(ptr: *const Type) -> Self {
        // there will be a compiler error if the passed in type doesn't exist in the enum
        Self {
            repr: TaggedPtr::init(ptr, Type::TAG),
            _types: PhantomData,
        }
    }

    #[inline]
    pub fn is_null(self) -> bool {
        self.repr.ptr_bits() == 0
    }
}
